File: go/test/webhooks_subscription_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/hubspot-webhooks-sdk/go"
	"github.com/voxgig-sdk/hubspot-webhooks-sdk/go/core"

	vs "github.com/voxgig-sdk/hubspot-webhooks-sdk/go/utility/struct"
)

func TestWebhooksSubscriptionEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.WebhooksSubscription(nil)
		if ent == nil {
			t.Fatal("expected non-nil WebhooksSubscriptionEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := webhooks_subscriptionBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "webhooks_subscription." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		webhooksSubscriptionRef01Ent := client.WebhooksSubscription(nil)
		webhooksSubscriptionRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "webhooks_subscription"}), "webhooks_subscription_ref01"))
		webhooksSubscriptionRef01Data["app_id"] = setup.idmap["app01"]

		webhooksSubscriptionRef01DataResult, err := webhooksSubscriptionRef01Ent.Create(webhooksSubscriptionRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		webhooksSubscriptionRef01Data = core.ToMapAny(entityData(webhooksSubscriptionRef01DataResult))
		if webhooksSubscriptionRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if webhooksSubscriptionRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		webhooksSubscriptionRef01DataUp0Up := map[string]any{
			"id": webhooksSubscriptionRef01Data["id"],
			"app_id": setup.idmap["app_id"],
		}

		webhooksSubscriptionRef01MarkdefUp0Name := "createdAt"
		webhooksSubscriptionRef01MarkdefUp0Value := fmt.Sprintf("Mark01-webhooks_subscription_ref01_%d", setup.now)
		webhooksSubscriptionRef01DataUp0Up[webhooksSubscriptionRef01MarkdefUp0Name] = webhooksSubscriptionRef01MarkdefUp0Value

		webhooksSubscriptionRef01ResdataUp0Result, err := webhooksSubscriptionRef01Ent.Update(webhooksSubscriptionRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		webhooksSubscriptionRef01ResdataUp0 := core.ToMapAny(entityData(webhooksSubscriptionRef01ResdataUp0Result))
		if webhooksSubscriptionRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if webhooksSubscriptionRef01ResdataUp0["id"] != webhooksSubscriptionRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if webhooksSubscriptionRef01ResdataUp0[webhooksSubscriptionRef01MarkdefUp0Name] != webhooksSubscriptionRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", webhooksSubscriptionRef01MarkdefUp0Name, webhooksSubscriptionRef01ResdataUp0[webhooksSubscriptionRef01MarkdefUp0Name])
		}

		// LOAD
		webhooksSubscriptionRef01MatchDt0 := map[string]any{
			"id": webhooksSubscriptionRef01Data["id"],
		}
		webhooksSubscriptionRef01DataDt0Loaded, err := webhooksSubscriptionRef01Ent.Load(webhooksSubscriptionRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		webhooksSubscriptionRef01DataDt0LoadResult := core.ToMapAny(entityData(webhooksSubscriptionRef01DataDt0Loaded))
		if webhooksSubscriptionRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if webhooksSubscriptionRef01DataDt0LoadResult["id"] != webhooksSubscriptionRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func webhooks_subscriptionBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "webhooks_subscription", "WebhooksSubscriptionTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read webhooks_subscription test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse webhooks_subscription test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"webhooks_subscription01", "webhooks_subscription02", "webhooks_subscription03", "2026_0901", "2026_0902", "2026_0903", "app01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID": idmap,
		"HUBSPOT_WEBHOOKS_TEST_LIVE":      "FALSE",
		"HUBSPOT_WEBHOOKS_TEST_EXPLAIN":   "FALSE",
		"HUBSPOT_WEBHOOKS_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add app_id alias for update test.
	if idmapResolved["app_id"] == nil {
		idmapResolved["app_id"] = idmapResolved["app01"]
	}

	if env["HUBSPOT_WEBHOOKS_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["HUBSPOT_WEBHOOKS_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewHubspotWebhooksSDK(core.ToMapAny(mergedOpts))
	}

	live := env["HUBSPOT_WEBHOOKS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["HUBSPOT_WEBHOOKS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/webhooks_subscription_entity_test.lua
-- WebhooksSubscription entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("hubspot-webhooks_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("WebhooksSubscriptionEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:WebhooksSubscription(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = webhooks_subscription_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "webhooks_subscription." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local webhooks_subscription_ref01_ent = client:WebhooksSubscription(nil)
    local webhooks_subscription_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.webhooks_subscription"), "webhooks_subscription_ref01"))
    webhooks_subscription_ref01_data["2026_09_id"] = setup.idmap["2026_0901"]

    local webhooks_subscription_ref01_data_result, err = webhooks_subscription_ref01_ent:create(webhooks_subscription_ref01_data, nil)
    assert.is_nil(err)
    webhooks_subscription_ref01_data = helpers.to_map(type(webhooks_subscription_ref01_data_result) == 'table' and webhooks_subscription_ref01_data_result.data_get and webhooks_subscription_ref01_data_result:data_get() or webhooks_subscription_ref01_data_result)
    assert.is_not_nil(webhooks_subscription_ref01_data)
    assert.is_not_nil(webhooks_subscription_ref01_data["id"])

    -- UPDATE
    local webhooks_subscription_ref01_data_up0_up = {
      id = webhooks_subscription_ref01_data["id"],
      ["2026_09_id"] = setup.idmap["2026_09_id"],
    }

    local webhooks_subscription_ref01_markdef_up0_name = "createdAt"
    local webhooks_subscription_ref01_markdef_up0_value = "Mark01-webhooks_subscription_ref01_" .. tostring(setup.now)
    webhooks_subscription_ref01_data_up0_up[webhooks_subscription_ref01_markdef_up0_name] = webhooks_subscription_ref01_markdef_up0_value

    local webhooks_subscription_ref01_resdata_up0_result, err = webhooks_subscription_ref01_ent:update(webhooks_subscription_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local webhooks_subscription_ref01_resdata_up0 = helpers.to_map(type(webhooks_subscription_ref01_resdata_up0_result) == 'table' and webhooks_subscription_ref01_resdata_up0_result.data_get and webhooks_subscription_ref01_resdata_up0_result:data_get() or webhooks_subscription_ref01_resdata_up0_result)
    assert.is_not_nil(webhooks_subscription_ref01_resdata_up0)
    assert.are.equal(webhooks_subscription_ref01_resdata_up0["id"], webhooks_subscription_ref01_data_up0_up["id"])
    assert.are.equal(webhooks_subscription_ref01_resdata_up0[webhooks_subscription_ref01_markdef_up0_name], webhooks_subscription_ref01_markdef_up0_value)

    -- LOAD
    local webhooks_subscription_ref01_match_dt0 = {
      id = webhooks_subscription_ref01_data["id"],
    }
    local webhooks_subscription_ref01_data_dt0_loaded, err = webhooks_subscription_ref01_ent:load(webhooks_subscription_ref01_match_dt0, nil)
    assert.is_nil(err)
    local webhooks_subscription_ref01_data_dt0_load_result = helpers.to_map(type(webhooks_subscription_ref01_data_dt0_loaded) == 'table' and webhooks_subscription_ref01_data_dt0_loaded.data_get and webhooks_subscription_ref01_data_dt0_loaded:data_get() or webhooks_subscription_ref01_data_dt0_loaded)
    assert.is_not_nil(webhooks_subscription_ref01_data_dt0_load_result)
    assert.are.equal(webhooks_subscription_ref01_data_dt0_load_result["id"], webhooks_subscription_ref01_data["id"])

  end)
end)

function webhooks_subscription_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/webhooks_subscription/WebhooksSubscriptionTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read webhooks_subscription test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "webhooks_subscription01", "webhooks_subscription02", "webhooks_subscription03", "2026_0901", "2026_0902", "2026_0903" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID"] = idmap,
    ["HUBSPOT_WEBHOOKS_TEST_LIVE"] = "FALSE",
    ["HUBSPOT_WEBHOOKS_TEST_EXPLAIN"] = "FALSE",
    ["HUBSPOT_WEBHOOKS_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["HUBSPOT_WEBHOOKS_TEST_WEBHOOKS_SUBSCRIPTION_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["2026_09_id"] == nil then
    idmap_resolved["2026_09_id"] = idmap_resolved["2026_0901"]
  end

  if env["HUBSPOT_WEBHOOKS_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["HUBSPOT_WEBHOOKS_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["HUBSPOT_WEBHOOKS_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["HUBSPOT_WEBHOOKS_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
